### PR TITLE
jordgubb input

### PR DESCRIPTION
**Pull request by FredTheDino**
Adds the basics for simple input-manangement - there's a lot more that can be done here, but I did this as simply as I could to make it easy to work with, it might not even be good enough to use TBH. But sound is next!

Some extensions:
 - Gamepad support
 - Check for `pressed` `held` `released`

### CODE REVIEW

File: jordgubb/input.sy [hunk 1]
```diff
@@ -0,0 +1,75 @@
+// Based on scancodes
+Key :: enum
+    A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z,
+    N1, N2, N3, N4, N5, N6, N7, N8, N9, N0,
+    Return, Escape, Backspace, Tab, Space,
+    LCtrl, LShift, LAlt, LGui,
+    RCtrl, RShift, RAlt, RGui,
+    Up, Down, Left, Right,
+end
```
Comment (EMattfolk) [inline jordgubb/input.sy:15]: :eyes:

**Review by EMattfolk — APPROVED** [review]

File: jordgubb/gfx.sy [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+from love use love
+
+Color :: blob {
+    red: float,
+    green: float,
+    blue: float,
+    alpha: float,
+}
+
+// NOTE(ed): Might be nice to convert HSL to RGB?
```
Comment (sornas) [inline jordgubb/gfx.sy:11]: :+1:

File: jordgubb/gfx.sy [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+end
+rgba :: fn r, g, b, a -> Color
+    Color { red: r, green: g, blue: b, alpha: a }
+end
+stableRandomColor :: fn i: float -> Color
+    r :: rem' 0.3 + i * 3.33, 1.
+    g :: rem' 0.2 + i * 7.77, 1.
+    b :: rem' 0.7 + i * 5.55, 1.
+    rgb' r, g, b
+end
```
Comment (sornas) [inline jordgubb/gfx.sy:23]: would look nicer with just a randomized hue in hsv. but good for now

File: jordgubb/gfx.sy [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+setColor :: fn color do
+    love.graphics.setColor' (color.red, color.green, color.blue, color.alpha)
+end
+
+drawLine :: fn color, a, b do
+    setColor' color
+    love.graphics.line' a[0], a[1], b[0], b[1]
+end
```
Comment (sornas) [inline jordgubb/gfx.sy:32]: patterns in parameters maybe? so we could write `fn color, (x0, y0), (x1, y1)`. not right now i realize but in the future

File: jordgubb/gfx.sy [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+drawRect :: fn color, p, dim do
+    setColor' color
+    love.graphics.rectangle' "fill", p[0], p[1], dim[0], dim[1]
+end
+
+drawRectCentered :: fn color, p, dim do
+    drawRect' color, p - dim / (2., 2.), dim
+end
+
+drawRectOutline :: fn color, p, dim do
```
Comment (sornas) [inline jordgubb/gfx.sy:53]:
```suggestion
drawRectCornersOutline :: fn color, p, dim do
```

File: jordgubb/gfx.sy [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    setColor' color
+    love.graphics.circle' "fill", p[0], p[1], r
+end
+
+drawCircleOutline :: fn color, p, r do
+    setColor' color
+    love.graphics.circle' "line", p[0], p[1], r
+end
+
+drawRect :: fn color, p, dim do
```
Comment (sornas) [inline jordgubb/gfx.sy:44]:
```suggestion
drawRectCorners :: fn color, p, dim do
```

File: jordgubb/gfx.sy [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    drawTextF: fn Color, str, (float, float), Alignment, float -> void,
+    drawTextFT: fn Color, str, (float, float), Alignment, float, float, float -> void,
+}
+
+drawText :: fn color, text, pos, r, s do
+    setColor' color
+    love.graphics.print' text, pos[0], pos[1], r, s, s
+end
+
+drawTextFT :: fn color, text, pos, alignment, maxWidth, r, s do
```
Comment (sornas) [inline jordgubb/gfx.sy:91]: i think `FT` is unclear what it does from the name

File: jordgubb/gfx.sy [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+    img: *,
+    getWidth: fn -> float,
+    getHeight: fn -> float,
+    draw: fn (float, float), float, (float, float) -> void,
+    drawTinted: fn Color, (float, float), float, (float, float) -> void,
+    drawCenterd: fn (float, float), float, (float, float) -> void,
+    drawTintedCentered: fn Color, (float, float), float, (float, float) -> void,
+}
+
+drawImage :: fn tint, img: l.ImageData, pos, rot, scale do
```
Comment (sornas) [inline jordgubb/gfx.sy:129]:
```suggestion
drawImageCornered :: fn tint, img: l.ImageData, pos, rot, scale do
```

File: jordgubb/gfx.sy [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+    w :: spriteDim[0]
+    h :: spriteDim[1]
+    quads :: []
+    // TODO(ed): Maybe remove the remained so we only get full tiles?
+    // TODO(ed): Maybe add a divisibility test here?
+    for_range' 0, as_int(img.getWidth()), w, fn x -> do
+        for_range' 0, as_int(img.getHeight()), h, fn y -> do
+            quads -> push' ((x, y), (w, h))
+        end
+    end
```
Comment (sornas) [inline jordgubb/gfx.sy:197]: remember when we couldn't have for loops in other for loops? good times. (i had to do a double take here xd)

File: jordgubb/gfx.sy [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+    // so feel free to add them if you need them:
+    // https://love2d.org/wiki/ParticleSystem
+
+    update: fn float -> void,
+    draw: fn Color -> void,
+    drawAt: fn Color, (float, float) -> void,
+}
+
+drawParticlesAt :: fn tint, particles: Particles, pos do
+end
```
Comment (sornas) [inline jordgubb/gfx.sy:245]: hmm

File: jordgubb/gfx.sy [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+    loop curr < hi do
+        f' curr
+        curr += step
+    end
+end
+
+// (w, h) - all in pixels
+newImplicitSpriteSheet :: fn img: Image, spriteDim: (int, int) ->
+    w :: spriteDim[0]
+    h :: spriteDim[1]
```
Comment (sornas) [inline jordgubb/gfx.sy:189]: NICE

File: jordgubb/love.sy [hunk 12]
```diff
@@ -10,22 +10,22 @@ ParticleSystem :: blob {
...
+    setPosition: fn *, float, float -> void,
+
     update: fn *, float -> void,
 }
 
 // Place holder before we get `newtype`
-ImageData :: blob { _001_unique_type: int, }
+ImageData :: blob {
+    _001_unique_type: int,
+    // Uncomment for crash
```
Comment (sornas) [inline jordgubb/love.sy:22]: ?

**Review by sornas — APPROVED** [review]

File: jordgubb/gfx.sy (hunk 4 above)
Comment (FredTheDino) [reply jordgubb/gfx.sy:32]: Have you added it to *the list*?

File: jordgubb/gfx.sy (hunk 5 above)
Comment (FredTheDino) [reply jordgubb/gfx.sy:53]: Not fond of this name change - I read `corners` as the `verticies`. I think the old name is clearer.

File: jordgubb/gfx.sy (hunk 7 above)
Comment (FredTheDino) [reply jordgubb/gfx.sy:91]: `FormattedTransformed`
It was too long for me to be bothered to write it - I felt the abbreviation was better.

**Pull request merged**